### PR TITLE
OPMPHM: The opmphmInit

### PR DESCRIPTION
**Pull request by KurtMi**
Changed the structs and renamed mapping to init to implement the first part of my OPMPHM.

@markus2330 please add me to Whilelist

### CODE REVIEW

File: src/libs/elektra/opmphm.c [hunk 1]
```diff
@@ -11,89 +11,137 @@
 #include <kdblogger.h>
 #include <kdbopmphm.h>
 
-#include <stdio.h>  //debug
+#include <stdio.h>  //pfdebug
```
Comment (markus2330) [inline src/libs/elektra/opmphm.c]: Seems to be a leftover?

File: src/libs/elektra/opmphm.c [hunk 2]
```diff
@@ -11,89 +11,137 @@
...
-		uint32_t h2 = edges[i].h[2];
-		fprintf (f, "	%u -- %u [label=\"%s\"]\n", h1, h2, fpOpmhpmGetString (data[i]));
-	}
-	fprintf (f, "}\n");
-	fclose (f);
-}
-
-/** The first phase of the order preserving minimal perfect hash map build. */
-int opmphmMapping (OPMPHM * opmphm, Vertex * vertices, Edge * edges, OPMPHMinit * init, size_t n)
+/** The order preserving minimal perfect hash map Init. */
```
Comment (markus2330) [inline src/libs/elektra/opmphm.c]: Redundant explanation, parameters are not explained.

File: src/libs/elektra/opmphm.c [hunk 3]
```diff
@@ -11,89 +11,137 @@
...
-	for (size_t i = 0; i < r; ++i)
+	// sort elements in sortOrder with radixsort
+	// determine the maximum bucket capacity
+	size_t numberOfElements = 1;
+	for (unsigned int t = 1; t < OPMPHMNOEIOP; ++t)
+	{
+		numberOfElements = numberOfElements * w;
+	}
+	ELEKTRA_LOG ("OPMPHM w= %lu n= %lu opmphmRatio= %f radix numberOfElements per Bucket= %lu", w, n, opmphmRatio, numberOfElements);
+	OpmphmOrder ** buckets = malloc (sizeof (OpmphmOrder *) * w * numberOfElements);
```
Comment (markus2330) [inline src/libs/elektra/opmphm.c]: please use `elektraMalloc` and `elektraFree`.

File: tests/ctest/test_opmphm_init.c [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+		{
+			for (size_t n = 0; n < getPower (w, OPMPHMNOEIOP); ++n)
+			{
+				OpmphmOrder o;
+				getOrderedPair (&o, w, n);
+				succeed_if (getNumber (&o, w) == n, "relation between Ordered Pair and Integers wrong");
+			}
+		}
+	}
+	/* test w, always full. w^OPMPHMNOEIOP elements in each run.
```
Comment (markus2330) [inline tests/ctest/test_opmphm_init.c]: Please add some explanation what w is.

File: src/include/kdbopmphm.h [hunk 5]
```diff
@@ -1,76 +1,52 @@
...
 #define OPMPHM_H
 
+#include <stdbool.h>
 #include <stdint.h>
 #include <stdlib.h>
 
 /**
- * Vertex represents the vertex of the bipartite graph.
- * Only needed for build, 2*r times (see opmphmR, below).
+ * NOEIOP stands for number of elements in ordered pair.
```
Comment (markus2330) [inline src/include/kdbopmphm.h]: Please a bit more information, why do you know the number of elements?

**Comment by sanssecours** [comment]
@KurtMi 👋 I tested your changes, as you requested. Issue #1444 seems to be fixed. I still think that the problem in issue #1444 is the logger code itself though, not your usage of the logging functionality.

**Comment by markus2330** [comment]
add to whitelist

**Comment by KurtMi** [comment]
jenkins build gcc-i386 please

**Comment by markus2330** [comment]
Thank you, great job! Looking forward to see more!

**Pull request merged**